### ISSUE TITLE
`GithubComment` creation is not possible when using it is created as part of `ObjectTemplate` along with Jinja2 templated field

### ISSUE THREAD

**Issue by madhukarmmallia-plivo** [kluctl/template-controller#58]
## Overview:
We are trying to setup an ephemeral env for ease of development using the resources available with template-controller. Our use case involves creating a `GithubComment` resource as part of the `ObjectTemplate` so that we can add a comment to the pull request. This would mean dynamically creating the `GithubComment` with the pull request ID fetched by `ListGithubPullRequests`. To achieve this, we will have to use jinja2 templating for `spec.github.pullRequestId` either with double-quotes (identified as a string) or without it (throws invalid map key error when trying to create resources). 

Template controller version used: v0.7.0

### Scenario 1: `spec.github.pullRequestId` as Jinja2 template with double quotes (`"{{ matrix.pr.number }}"` identified as string)
```
---
apiVersion: templates.kluctl.io/v1alpha1
kind: ListGithubPullRequests
metadata:
  name: demo-service
  namespace: demo
spec:
  base: dev
  head: demo-org:.*
  interval: 1m
  limit: 5
  owner: demo-org
  repo: demo-service
  state: open
  tokenRef:
    key: GITHUB_TOKEN
    secretName: demo-service
---
apiVersion: templates.kluctl.io/v1alpha1
kind: ObjectTemplate
metadata:
  name: demo-service
  namespace: demo
spec:
  serviceAccountName: demo-preview-env
  prune: true
  interval: 1m
  matrix:
    - name: pr
      object:
        ref:
          apiVersion: templates.kluctl.io/v1alpha1
          kind: ListGithubPullRequests
          name: demo-service
        jsonPath: status.pullRequests
        expandLists: true
  templates:
    - object:
        apiVersion: templates.kluctl.io/v1alpha1
        kind: GithubComment
        metadata:
          name: demo-service-{{ matrix.pr.number }}
          namespace: demo
        spec:
          github:
            owner: demo-org
            repo: demo-service
            pullRequestId: "{{ matrix.pr.number }}"
            tokenRef:
              secretName: demo-service
              key: "GITHUB_TOKEN"
          comment:
            source:
              text: "Demo comment"
```

#### Error:
```
1 error occurred:\n\t* failed to create typed patch object (demo/demo-service-1; templates.kluctl.io/v1alpha1, Kind=GithubComment): .spec.github.pullRequestId: expected numeric (int or float), got string\n\n
```

### Scenario 2: `spec.github.pullRequestId` as Jinja2 template without double quotes (`{{ matrix.pr.number }}`)
```
---
apiVersion: templates.kluctl.io/v1alpha1
kind: ListGithubPullRequests
metadata:
  name: demo-service
  namespace: demo
spec:
  base: dev
  head: demo-org:.*
  interval: 1m
  limit: 5
  owner: demo-org
  repo: demo-service
  state: open
  tokenRef:
    key: GITHUB_TOKEN
    secretName: demo-service
---
apiVersion: templates.kluctl.io/v1alpha1
kind: ObjectTemplate
metadata:
  name: demo-service
  namespace: demo
spec:
  serviceAccountName: demo-preview-env
  prune: true
  interval: 1m
  matrix:
    - name: pr
      object:
        ref:
          apiVersion: templates.kluctl.io/v1alpha1
          kind: ListGithubPullRequests
          name: demo-service
        jsonPath: status.pullRequests
        expandLists: true
  templates:
    - object:
        apiVersion: templates.kluctl.io/v1alpha1
        kind: GithubComment
        metadata:
          name: demo-service-{{ matrix.pr.number }}
          namespace: demo
        spec:
          github:
            owner: demo-org
            repo: demo-service
            pullRequestId: {{ matrix.pr.number }}
            tokenRef:
              secretName: demo-service
              key: "GITHUB_TOKEN"
          comment:
            source:
              text: "Demo comment"
```

#### Error:

```
error: error parsing github-comment-test.yaml: error converting YAML to JSON: yaml: invalid map key: map[interface {}]interface {}{"matrix.pr.number":interface {}(nil)}
```

## Suggested Solution:

Currently, the field `spec.github.pullRequestId` accepts only `int`. To handle the above scenario, my suggestion is to change this to type `IntOrString` [k8s intstr](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/intstr#IntOrString) type. I can work on this and raise a PR if the solutions looks good.

**Comment by madhukarmmallia-plivo** [comment]
https://github.com/kluctl/template-controller/pull/59 contains the changes suggested for handling this issue.

**Comment by codablock** [comment]
@madhukarmmallia-plivo Thanks, I have merged #59 now.

However, I believe that this topic is something that will constantly re-occur, as other places in the k8s world will have the same issue.

The problem here is, that `template-controller` treats the field as string when it initially parses the yaml. Then, after rendering the templates inside the object template, it can't know if a value that ends up being a number should be treated as a number or as a string now. Right now, it always defaults to emitting a string when sending the resulting json to the k8s api, which then causes failures. This issue is present in all kinds of k8s resources that we have no control of (so we can't introduce IntOrString there).

The solution that always works is to use `raw` instead of `object` in `templates`. This treats the whole object as a string and thus allows to handle all cases in yaml/json parsing.
